### ISSUE TITLE
single_char_pattern with constants

### ISSUE THREAD

**Issue by HongxuChen** [rust-lang-nursery/rust-clippy#3204]
cargo-clippy can successfully detect single_char_pattern when the char is defined as a constant rather than a literal. However the hint message is wrongly handled by manipulating the char variable rather than representing char value,  which can even trigger bounding error.
```Rust
# src/main.rs
const SS:&str = "#";
const S:&str = "#";
pub fn main() {
    let line = String::from("a#b");
    if let Some(pos) = line.find("#") {
        println!("literal: pos={}", pos);
    }
    if let Some(pos) = line.find(SS) {
        println!("SS: pos={}", pos);
    }
    if let Some(pos) = line.find(S) {
        println!("S: pos={}", pos);
    }
}
```

The clippy report may be like:

```
~/test-proj #3045  cargo clippy
    Checking test-proj v0.1.0 (/home/hongxu/test-proj)
warning: single-character string constant used as pattern
 --> src/main.rs:6:34
  |
6 |     if let Some(pos) = line.find("#") {
  |                                  ^^^ help: try using a char instead: `'#'`
  |
  = note: #[warn(clippy::single_char_pattern)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#single_char_pattern
warning: single-character string constant used as pattern
 --> src/main.rs:9:34
  |
9 |     if let Some(pos) = line.find(SS) {
  |                                  ^^ help: try using a char instead: `''`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#single_char_pattern
thread 'main' panicked at 'begin <= end (1 <= 0) when slicing `S`', libcore/str/mod.rs:2098:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: rustc 1.30.0-nightly (20dc0c507 2018-09-19) running on x86_64-unknown-linux-gnu
note: compiler flags: -C debuginfo=2 -C target-cpu=native --crate-type bin
note: some of the compiler flags provided by cargo are hidden
error: Could not compile `test-proj`.
To learn more, run the command again with --verbose.
```

**Comment by mikerite** [comment]
IMO, this shouldn't lint at all. Replacing a constant with a char literal seems like questionable advice.
